cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,1111 @@
+	zvol_io_hdr_t hdr = {0, };
+	char *name = NULL;
+	char *buf = NULL;
+
+
+	sfd = create_and_bind(mgmt_port);
+	if (sfd == -1) {
+		goto exit;
+	}
+
+	buf = get_controller_ip_address();
+	if (buf == NULL) {
+		printf("parsing IP address did not work\n");
+		goto exit;
+	}
+
+	printf("Controller IP address is: %s", buf);
+	bzero((char *)&istgt_addr, sizeof (istgt_addr));
+	istgt_addr.sin_family = AF_INET;
+	istgt_addr.sin_addr.s_addr = inet_addr(buf);
+	istgt_addr.sin_port = htons(6060);
+	free(buf);
+retry:
+	rc = connect(sfd, (struct sockaddr *)&istgt_addr, sizeof (istgt_addr));
+	if ((rc == -1) && ((errno == EINTR) || (errno == ECONNREFUSED) ||
+	    (errno == ETIMEDOUT) || (errno == EINPROGRESS))) {
+		ZREPL_ERRLOG("Failed to connect to istgt_controller"
+		    " with err:%d\n", errno);
+		sleep(2);
+		printf("Retrying ....\n");
+		goto retry;
+	} else {
+		printf("Connection to TGT controller successful\n");
+		ZREPL_LOG("Connection to TGT controller iss successful\n");
+	}
+
+	while (1) {
+		bzero(&hdr, sizeof (hdr));
+		count = read(sfd, (char *)&hdr, sizeof (hdr));
+		if (count <= 0) {
+			ZREPL_ERRLOG("Replica-iSCSI Tgt connection got "
+			    "disconnected with err:%d\n", errno);
+			/*
+			 * Error has occurred on this socket
+			 * close it and open a new socket after
+			 * 5 sec of sleep.
+			 */
+			close(sfd);
+			printf("Retrying ....\n");
+			sleep(5);
+			sfd = create_and_bind(mgmt_port);
+			if (sfd == -1) {
+				goto exit;
+			}
+
+retry1:
+			rc = connect(sfd, (struct sockaddr *)&istgt_addr,
+			    sizeof (istgt_addr));
+			if ((rc == -1) && ((errno == EINTR) ||
+			    (errno == ECONNREFUSED) || (errno == ETIMEDOUT))) {
+				ZREPL_ERRLOG("Failed to connect to"
+				    " istgt_controller with err:%d\n",
+				    errno);
+				sleep(2);
+				goto retry1;
+			} else {
+				printf("Connection to TGT controller "
+				    "successful\n");
+				ZREPL_LOG("Connection to TGT controller"
+				    "is successful\n");
+			}
+			continue;
+		}
+
+		if (hdr.opcode == ZVOL_OPCODE_HANDSHAKE) {
+			name = kmem_alloc(
+			    hdr.len * sizeof (char), KM_SLEEP);
+			count = read(sfd, name, sizeof (char) * hdr.len);
+			if (count == -1) {
+				ZREPL_ERRLOG("Read from socket failed"
+				    " with err:%d\n", errno);
+				goto exit;
+			}
+
+			rc = uzfs_zvol_mgmt_do_handshake(&hdr, sfd, name);
+			free(name);
+			if (rc == -1) {
+				ZREPL_ERRLOG("handshake failed with"
+				    " errno:%d\n", errno);
+				goto exit;
+			}
+		}
+	}
+exit:
+	printf("uzfs_zvol_mgmt_thread thread exiting\n");
+	zk_thread_exit();
+}
+#if 0
+/*
+ * One thread per replica, which will be
+ * responsible for initial handshake and
+ * exchanging info like IP add, port etc.
+ */
+static void
+uzfs_zvol_mgmt_thread(void *arg)
+{
+
+	int sfd, efd, rc, count;
+	struct epoll_event event;
+	struct sockaddr_in istgt_addr;
+	zvol_io_hdr_t hdr = {0, };
+	struct epoll_event *events = NULL;
+	char *name = NULL;
+	char *buf = NULL;
+
+
+	sfd = create_and_bind(mgmt_port);
```

Comment:
'bind' is not required.

Reply:
I can understand we do not need to explicitly bind this since connect anyway bind socket before establishing connection to iSCSI controller.